File: packages/neptune_flutter_ui/lib/src/widgets/neptune_money_movement.dart
// © 2026 Neptune.Fintech (neptune.ly) · Neptune Odyssey Community License v1.0

import 'package:flutter/material.dart';

import '../theme/accessibility.dart';
import '../theme/extensions.dart';
import '../theme/neptune_theme.dart';

/// A horizontal progress stepper (web `<npt-stepper>`): numbered nodes joined by
/// connector lines, with a label under each. Done steps are filled
/// [ColorScheme.primary], the active step is outlined primary, and future steps
/// use [ColorScheme.outlineVariant]. Spoken as one stop - "step 2 of 4,
/// Review, 1 of 4 steps done" - and re-announced as the customer advances.
/// Theme-only, RTL-safe.
class NeptuneStepper extends StatelessWidget {
  final List<String> steps;
  final int active;

  const NeptuneStepper({
    super.key,
    required this.steps,
    required this.active,
  });

  @override
  Widget build(BuildContext context) {
    final scheme = Theme.of(context).colorScheme;
    final shape = Theme.of(context).extension<NptShape>()!;
    final type = Theme.of(context).extension<NptType>()!;
    final text = Theme.of(context).textTheme;
    final strings = NeptuneAccessibility.of(context);

    final children = <Widget>[];
    for (var i = 0; i < steps.length; i++) {
      final state = i < active
          ? _StepState.done
          : i == active
              ? _StepState.active
              : _StepState.upcoming;

      // Node colours by state.
      final Color fill;
      final Color border;
      final Color fg;
      switch (state) {
        case _StepState.done:
        case _StepState.active:
          fill = scheme.primary;
          border = scheme.primary;
          fg = scheme.onPrimary;
          break;
        case _StepState.upcoming:
          fill = scheme.surfaceContainerHighest;
          border = scheme.outlineVariant;
          fg = scheme.onSurfaceVariant;
          break;
      }
      // The active node reads as an outlined ring (primary outline, hollow fill).
      final isActive = state == _StepState.active;

      final node = Container(
        width: 32,
        height: 32,
        alignment: Alignment.center,
        decoration: BoxDecoration(
          color: isActive ? scheme.surface : fill,
          shape: BoxShape.circle,
          border: Border.all(color: border, width: 2),
        ),
        child: state == _StepState.done
            ? Icon(Icons.check, size: 18, color: fg)
            : Text(
                '${i + 1}',
                style: text.labelLarge?.copyWith(
                  fontFamily: type.num,
                  color: isActive ? scheme.primary : fg,
                  fontWeight: FontWeight.w600,
                ),
              ),
      );

      final label = SizedBox(
        width: 80,
        child: Text(
          steps[i],
          textAlign: TextAlign.center,
          maxLines: 2,
          overflow: TextOverflow.ellipsis,
          style: text.labelSmall?.copyWith(
            color: isActive ? scheme.onSurface : scheme.onSurfaceVariant,
            fontWeight: isActive ? FontWeight.w600 : FontWeight.w500,
          ),
        ),
      );

      children.add(
        Column(
          mainAxisSize: MainAxisSize.min,
          children: [node, const SizedBox(height: 8), label],
        ),
      );

      // Connector between this node and the next.
      if (i < steps.length - 1) {
        final done = i < active;
        children.add(
          Expanded(
            child: Container(
              height: 2,
              margin: const EdgeInsetsDirectional.only(top: 15, start: 4, end: 4),
              decoration: BoxDecoration(
                color: done ? scheme.primary : scheme.outlineVariant,
                borderRadius: shape.rXs,
              ),
            ),
          ),
        );
      }
    }

    final current = active.clamp(0, steps.isEmpty ? 0 : steps.length - 1);
    final spoken = steps.isEmpty
        ? ''
        : '${strings.step(current + 1, steps.length, steps[current])}, '
            '${strings.stepsProgress(active.clamp(0, steps.length), steps.length)}';

    return Semantics(
      liveRegion: true,
      label: spoken,
      excludeSemantics: true,
      child: Row(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: children,
      ),
    );
  }
}

enum _StepState { done, active, upcoming }

/// A transfer review card (web `<npt-transfer-review>`): a
/// [ColorScheme.surfaceContainer] card with key/value rows (From, To, Amount,
/// optional Fee) and a highlighted, bold [Total] rendered in the money style.
/// Theme-only, RTL-safe.
class NeptuneTransferReview extends StatelessWidget {
  final String fromLabel;
  final String toLabel;
  final String amount;
  final String? fee;
  final String? total;
  final String? currency;

  /// Row captions. Default to the host's vocabulary
  /// ([NeptuneA11yStrings.from] etc.) so an Arabic app never shows "From".
  final String? fromCaption;
  final String? toCaption;
  final String? amountCaption;
  final String? feeCaption;
  final String? totalCaption;

  const NeptuneTransferReview({
    super.key,
    required this.fromLabel,
    required this.toLabel,
    required this.amount,
    this.fee,
    this.total,
    this.currency,
    this.fromCaption,
    this.toCaption,
    this.amountCaption,
    this.feeCaption,
    this.totalCaption,
  });

  @override
  Widget build(BuildContext context) {
    final scheme = Theme.of(context).colorScheme;
    final shape = Theme.of(context).extension<NptShape>()!;
    final text = Theme.of(context).textTheme;
    final strings = NeptuneAccessibility.of(context);
    final totalStyle = NeptuneTheme.moneyStyle(context, base: text.titleLarge)
        .copyWith(color: scheme.primary, fontWeight: FontWeight.w700);

    final rows = <Widget>[
      _ReviewRow(label: fromCaption ?? strings.from, value: fromLabel),
      const SizedBox(height: 12),
      _ReviewRow(label: toCaption ?? strings.to, value: toLabel),
      const SizedBox(height: 12),
      _ReviewRow(
        label: amountCaption ?? strings.amount,
        value: amount,
        numeric: true,
        currency: currency,
      ),
      if (fee != null) ...[
        const SizedBox(height: 12),
        _ReviewRow(
          label: feeCaption ?? strings.fee,
          value: fee!,
          numeric: true,
          currency: currency,
        ),
      ],
    ];

    final totalLabel = totalCaption ?? strings.total;

    return Container(
      decoration: BoxDecoration(
        color: scheme.surfaceContainer,
        borderRadius: shape.rLg,
      ),
      padding: const EdgeInsetsDirectional.all(20),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        mainAxisSize: MainAxisSize.min,
        children: [
          ...rows,
          if (total != null) ...[
            const SizedBox(height: 16),
            Divider(height: 1, thickness: 1, color: scheme.outlineVariant),
            const SizedBox(height: 16),
            // The total is the one number a customer must not mishear:
            // "Total: 1,251.500 Libyan dinars", one stop, currency spoken.
            Semantics(
              label:
                  '$totalLabel: ${NeptuneAccessibility.money(context, total!, currency: currency)}',
              excludeSemantics: true,
              child: Row(
              crossAxisAlignment: CrossAxisAlignment.baseline,
              textBaseline: TextBaseline.alphabetic,
              children: [
                Expanded(
                  child: Text(
                    totalLabel,
                    style: text.labelLarge?.copyWith(
                      color: scheme.onSurface,
                      fontWeight: FontWeight.w600,
                    ),
                  ),
                ),
                const SizedBox(width: 16),
                Flexible(
                  child: Row(
                    mainAxisSize: MainAxisSize.min,
                    crossAxisAlignment: CrossAxisAlignment.baseline,
                    textBaseline: TextBaseline.alphabetic,
                    children: [
                      Flexible(
                        child: Text(
                          total!,
                          maxLines: 1,
                          overflow: TextOverflow.ellipsis,
                          style: totalStyle,
                        ),
                      ),
                      if (currency != null) ...[
                        const SizedBox(width: 4),
                        Text(
                          currency!,
                          style: text.bodyMedium?.copyWith(
                            color: scheme.primary.withValues(alpha: 0.85),
                          ),
                        ),
                      ],
                    ],
                  ),
                ),
              ],
              ),
            ),
          ],
        ],
      ),
    );
  }
}

/// One key/value line in a [NeptuneTransferReview]. Numeric values use the money
/// style for column-aligned figures and are spoken as money.
class _ReviewRow extends StatelessWidget {
  final String label;
  final String value;
  final bool numeric;
  final String? currency;

  const _ReviewRow({
    required this.label,
    required this.value,
    this.numeric = false,
    this.currency,
  });

  @override
  Widget build(BuildContext context) {
    final scheme = Theme.of(context).colorScheme;
    final text = Theme.of(context).textTheme;
    final valueStyle = numeric
        ? NeptuneTheme.moneyStyle(context, base: text.bodyLarge)
            .copyWith(color: scheme.onSurface)
        : text.bodyLarge?.copyWith(color: scheme.onSurface);
    final spokenValue = numeric
        ? NeptuneAccessibility.money(context, value, currency: currency)
        : value;

    return Semantics(
      label: '$label: $spokenValue',
      excludeSemantics: true,
      child: Row(
      crossAxisAlignment: CrossAxisAlignment.baseline,
      textBaseline: TextBaseline.alphabetic,
      children: [
        Text(
          label,
          style: text.bodyMedium?.copyWith(color: scheme.onSurfaceVariant),
        ),
        const SizedBox(width: 16),
        Expanded(
          child: Text(
            value,
            textAlign: TextAlign.end,
            style: valueStyle,
          ),
        ),
      ],
      ),
    );
  }
}

/// A selectable payment-method row (web `<npt-method-row>`): a leading rounded
/// icon, a title with optional subtitle, and a trailing radio that fills with a
/// check when [selected]. The whole row is tappable and at least 48dp tall.
/// Theme-only, RTL-safe.
class NeptuneMethodRow extends StatelessWidget {
  final IconData icon;
  final String title;
  final String? subtitle;
  final bool selected;
  final VoidCallback? onTap;

  const NeptuneMethodRow({
    super.key,
    required this.icon,
    required this.title,
    this.subtitle,
    this.selected = false,
    this.onTap,
  });

  @override
  Widget build(BuildContext context) {
    final scheme = Theme.of(context).colorScheme;
    final shape = Theme.of(context).extension<NptShape>()!;
    final text = Theme.of(context).textTheme;

    return Material(
      color: scheme.surfaceContainerLowest,
      borderRadius: shape.rMd,
      clipBehavior: Clip.antiAlias,
      // A radio-style choice: named, with its selected state, one stop.
      child: Semantics(
        button: onTap != null,
        inMutuallyExclusiveGroup: true,
        checked: selected,
        label: subtitle == null ? title : '$title, $subtitle',
        excludeSemantics: true,
        child: InkWell(
        onTap: onTap,
        child: Container(
          constraints: const BoxConstraints(minHeight: 64),
          decoration: BoxDecoration(
            borderRadius: shape.rMd,
            border: Border.all(
              color: selected ? scheme.primary : scheme.outlineVariant,
              width: selected ? 2 : 1,
            ),
          ),
          padding: const EdgeInsetsDirectional.symmetric(
            horizontal: 16,
            vertical: 12,
          ),
          child: Row(
            children: [
              Container(
                width: 40,
                height: 40,
                alignment: Alignment.center,
                decoration: BoxDecoration(
                  color: scheme.secondaryContainer,
                  borderRadius: shape.rSm,
                ),
                child: Icon(icon, size: 20, color: scheme.onSecondaryContainer),
              ),
              const SizedBox(width: 16),
              Expanded(
                child: Column(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  mainAxisSize: MainAxisSize.min,
                  children: [
                    Text(
                      title,
                      maxLines: 1,
                      overflow: TextOverflow.ellipsis,
                      style: text.bodyLarge?.copyWith(
                        color: scheme.onSurface,
                        fontWeight: FontWeight.w600,
                      ),
                    ),
                    if (subtitle != null) ...[
                      const SizedBox(height: 2),
                      Text(
                        subtitle!,
                        maxLines: 1,
                        overflow: TextOverflow.ellipsis,
                        style: text.bodyMedium
                            ?.copyWith(color: scheme.onSurfaceVariant),
                      ),
                    ],
                  ],
                ),
              ),
              const SizedBox(width: 12),
              _SelectIndicator(selected: selected),
            ],
          ),
        ),
        ),
      ),
    );
  }
}

/// A trailing radio that fills [ColorScheme.primary] with a check when selected,
/// or shows an empty outlined circle when not.
class _SelectIndicator extends StatelessWidget {
  final bool selected;

  const _SelectIndicator({required this.selected});

  @override
  Widget build(BuildContext context) {
    final scheme = Theme.of(context).colorScheme;

    return Container(
      width: 24,
      height: 24,
      alignment: Alignment.center,
      decoration: BoxDecoration(
        color: selected ? scheme.primary : null,
        shape: BoxShape.circle,
        border: Border.all(
          color: selected ? scheme.primary : scheme.outline,
          width: 2,
        ),
      ),
      child: selected
          ? Icon(Icons.check, size: 16, color: scheme.onPrimary)
          : null,
    );
  }
}

/// A selectable beneficiary tile (web `<npt-beneficiary-tile>`): a circular
/// avatar (provided [avatar] or generated initials), the [name], and a masked
/// [account] in tabular figures. When [selected] the tile lifts onto a
/// highlighted container with a primary ring. At least 48dp tall.
/// Theme-only, RTL-safe.
class NeptuneBeneficiaryTile extends StatelessWidget {
  final String name;
  final String? account;
  final Widget? avatar;
  final bool selected;
  final VoidCallback? onTap;

  const NeptuneBeneficiaryTile({
    super.key,
    required this.name,
    this.account,
    this.avatar,
    this.selected = false,
    this.onTap,
  });

  String _initials(String value) {
    final parts =
        value.trim().split(RegExp(r'\s+')).where((p) => p.isNotEmpty).toList();
    if (parts.isEmpty) return '•';
    final first = parts.first.characters.first;
    final last = parts.length > 1 ? parts.last.characters.first : '';
    final out = (first + last).toUpperCase();
    return out.isEmpty ? '•' : out;
  }

  @override
  Widget build(BuildContext context) {
    final scheme = Theme.of(context).colorScheme;
    final shape = Theme.of(context).extension<NptShape>()!;
    final type = Theme.of(context).extension<NptType>()!;
    final text = Theme.of(context).textTheme;

    final accountStyle = NeptuneTheme.moneyStyle(context, base: text.bodyMedium)
        .copyWith(color: scheme.onSurfaceVariant);

    // "Ahmed Ali, ending in 4 8 2 1, selected" - the check glyph is the only
    // visible selection cue and would otherwise be silent.
    final spoken = [
      name,
      if (account != null) NeptuneAccessibility.maskedNumber(context, account!),
    ].join(', ');

    return Material(
      color: selected ? scheme.surfaceContainerHigh : scheme.surface,
      borderRadius: shape.rMd,
      clipBehavior: Clip.antiAlias,
      child: Semantics(
        button: onTap != null,
        selected: selected,
        label: spoken,
        excludeSemantics: true,
        child: InkWell(
        onTap: onTap,
        child: Container(
          constraints: const BoxConstraints(minHeight: 56),
          decoration: BoxDecoration(
            borderRadius: shape.rMd,
            border: Border.all(
              color: selected ? scheme.primary : scheme.outlineVariant,
              width: selected ? 2 : 1,
            ),
          ),
          padding: const EdgeInsetsDirectional.symmetric(
            horizontal: 12,
            vertical: 12,
          ),
          child: Row(
            children: [
              SizedBox(
                width: 44,
                height: 44,
                child: avatar ??
                    Container(
                      alignment: Alignment.center,
                      decoration: BoxDecoration(
                        color: scheme.primaryContainer,
                        shape: BoxShape.circle,
                      ),
                      child: Text(
                        _initials(name),
                        style: text.labelLarge?.copyWith(
                          fontFamily: type.display,
                          color: scheme.onPrimaryContainer,
                          fontWeight: FontWeight.w600,
                        ),
                      ),
                    ),
              ),
              const SizedBox(width: 16),
              Expanded(
                child: Column(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  mainAxisSize: MainAxisSize.min,
                  children: [
                    Text(
                      name,
                      maxLines: 1,
                      overflow: TextOverflow.ellipsis,
                      style: text.bodyLarge?.copyWith(color: scheme.onSurface),
                    ),
                    if (account != null) ...[
                      const SizedBox(height: 2),
                      Text(
                        account!,
                        maxLines: 1,
                        overflow: TextOverflow.ellipsis,
                        style: accountStyle,
                      ),
                    ],
                  ],
                ),
              ),
              if (selected) ...[
                const SizedBox(width: 12),
                Icon(Icons.check_circle, size: 22, color: scheme.primary),
              ],
            ],
          ),
        ),
        ),
      ),
    );
  }
}

File: packages/neptune_kmp_ui/odyssey-compose-ui/src/commonMain/kotlin/ly/neptune/odyssey/ui/components/NeptuneMoneyMovement.kt
// © 2026 Neptune.Fintech (neptune.ly) · Neptune Odyssey Community License v1.0
//
// The transfer-flow building blocks: the progress stepper, the transfer
// review summary card, the selectable payment-method row and the selectable
// beneficiary tile. Web counterparts: `<npt-stepper>` / `<npt-transfer-review>`
// / `<npt-method-row>` / `<npt-beneficiary-tile>` · Flutter:
// neptune_money_movement.dart. Theme-only, RTL-safe, ≥48dp targets.

package ly.neptune.odyssey.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.material3.Icon
import androidx.compose.material3.LocalContentColor
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.hapticfeedback.HapticFeedback
import androidx.compose.ui.platform.LocalHapticFeedback
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import ly.neptune.odyssey.ui.glyphs.NptGlyphs
import ly.neptune.odyssey.ui.glyphs.NptStatusGlyphs
import ly.neptune.odyssey.ui.theme.NeptuneTheme
import ly.neptune.odyssey.ui.theme.NptFeedbackCue
import ly.neptune.odyssey.ui.theme.rememberNeptuneFontFamily

/**
 * A horizontal progress stepper: numbered nodes joined by connector lines,
 * with a label under each. Done steps are filled `primary` (check glyph),
 * the active step is an outlined primary ring on `surface`, and future steps
 * use `surfaceContainerHighest` / `outlineVariant`.
 *
 * Web counterpart: `<npt-stepper>` · Flutter: `NeptuneStepper`.
 */
@Composable
public fun NeptuneStepper(
    steps: List<String>,
    active: Int,
    modifier: Modifier = Modifier,
) {
    val scheme = MaterialTheme.colorScheme
    val typography = MaterialTheme.typography
    val shape = NeptuneTheme.shape

    Row(modifier, verticalAlignment = Alignment.Top) {
        steps.forEachIndexed { i, step ->
            val done = i < active
            val isActive = i == active

            // Node colours by state; the active node reads as an outlined
            // ring (primary outline, hollow surface fill).
            val fill = if (done || isActive) scheme.primary else scheme.surfaceContainerHighest
            val borderColor = if (done || isActive) scheme.primary else scheme.outlineVariant
            val fg = if (done || isActive) scheme.onPrimary else scheme.onSurfaceVariant

            Column(horizontalAlignment = Alignment.CenterHorizontally) {
                Box(
                    modifier = Modifier
                        .size(32.dp)
                        .clip(shape.rFull)
                        .background(if (isActive) scheme.surface else fill)
                        .border(2.dp, borderColor, shape.rFull),
                    contentAlignment = Alignment.Center,
                ) {
                    if (done) {
                        Icon(
                            NptGlyphs.check,
                            contentDescription = null,
                            tint = fg,
                            modifier = Modifier.size(18.dp),
                        )
                    } else {
                        Text(
                            text = NeptuneTheme.formatDigits("${i + 1}"),
                            style = NeptuneTheme.moneyStyle(base = typography.labelLarge).copy(
                                color = if (isActive) scheme.primary else fg,
                                fontWeight = FontWeight.W600,
                            ),
                        )
                    }
                }
                Spacer(Modifier.height(8.dp))
                Text(
                    text = step,
                    textAlign = TextAlign.Center,
                    maxLines = 2,
                    overflow = TextOverflow.Ellipsis,
                    style = typography.labelSmall.copy(
                        color = if (isActive) scheme.onSurface else scheme.onSurfaceVariant,
                        fontWeight = if (isActive) FontWeight.W600 else FontWeight.W500,
                    ),
                    modifier = Modifier.width(80.dp),
                )
            }

            // Connector between this node and the next.
            if (i < steps.lastIndex) {
                Box(
                    Modifier
                        .weight(1f)
                        .padding(top = 15.dp, start = 4.dp, end = 4.dp)
                        .height(2.dp)
                        .clip(shape.rXs)
                        .background(if (done) scheme.primary else scheme.outlineVariant),
                )
            }
        }
    }
}

/**
 * A transfer review card: a `surfaceContainer` card with key/value rows
 * (From, To, Amount, optional Fee) and a highlighted, bold Total rendered in
 * the money style with an optional trailing [currency] code.
 *
 * Web counterpart: `<npt-transfer-review>` · Flutter: `NeptuneTransferReview`.
 */
@Composable
public fun NeptuneTransferReview(
    fromLabel: String,
    toLabel: String,
    amount: String,
    modifier: Modifier = Modifier,
    fee: String? = null,
    total: String? = null,
    currency: String? = null,
) {
    val scheme = MaterialTheme.colorScheme
    val typography = MaterialTheme.typography
    val totalStyle = NeptuneTheme.moneyStyle(base = typography.titleLarge)
        .copy(color = scheme.primary, fontWeight = FontWeight.W700)

    Column(
        modifier = modifier
            .clip(NeptuneTheme.shape.rLg)
            .background(scheme.surfaceContainer)
            .padding(20.dp),
    ) {
        ReviewRow(label = "From", value = fromLabel)
        Spacer(Modifier.height(12.dp))
        ReviewRow(label = "To", value = toLabel)
        Spacer(Modifier.height(12.dp))
        ReviewRow(label = "Amount", value = amount, numeric = true)
        if (fee != null) {
            Spacer(Modifier.height(12.dp))
            ReviewRow(label = "Fee", value = fee, numeric = true)
        }
        if (total != null) {
            Spacer(Modifier.height(16.dp))
            Box(Modifier.fillMaxWidth().height(1.dp).background(scheme.outlineVariant))
            Spacer(Modifier.height(16.dp))
            Row {
                Text(
                    text = "Total",
                    style = typography.labelLarge.copy(
                        color = scheme.onSurface,
                        fontWeight = FontWeight.W600,
                    ),
                    modifier = Modifier.weight(1f).alignByBaseline(),
                )
                Spacer(Modifier.width(16.dp))
                Text(
                    text = NeptuneTheme.formatDigits(total),
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                    style = totalStyle,
                    modifier = Modifier.weight(1f, fill = false).alignByBaseline(),
                )
                if (currency != null) {
                    Spacer(Modifier.width(4.dp))
                    Text(
                        text = currency,
                        style = typography.bodyMedium.copy(
                            color = scheme.primary.copy(alpha = 0.85f),
                        ),
                        modifier = Modifier.alignByBaseline(),
                    )
                }
            }
        }
    }
}

/** One key/value line in a [NeptuneTransferReview]. Numeric values use the
 * money style for column-aligned figures. */
@Composable
private fun ReviewRow(
    label: String,
    value: String,
    numeric: Boolean = false,
) {
    val scheme = MaterialTheme.colorScheme
    val typography = MaterialTheme.typography
    val valueStyle = if (numeric) {
        NeptuneTheme.moneyStyle(base = typography.bodyLarge).copy(color = scheme.onSurface)
    } else {
        typography.bodyLarge.copy(color = scheme.onSurface)
    }

    Row {
        Text(
            text = label,
            style = typography.bodyMedium.copy(color = scheme.onSurfaceVariant),
            modifier = Modifier.alignByBaseline(),
        )
        Spacer(Modifier.width(16.dp))
        Text(
            text = if (numeric) NeptuneTheme.formatDigits(value) else value,
            textAlign = TextAlign.End,
            style = valueStyle,
            modifier = Modifier.weight(1f).alignByBaseline(),
        )
    }
}

/**
 * A selectable payment-method row: a leading rounded 24dp [icon] slot on a
 * `secondaryContainer` tile, a title with optional subtitle, and a trailing
 * radio that fills `primary` with a check when [selected]. The whole row is
 * tappable and at least 64dp tall.
 *
 * Web counterpart: `<npt-method-row>` · Flutter: `NeptuneMethodRow`.
 */
@Composable
public fun NeptuneMethodRow(
    icon: @Composable () -> Unit,
    title: String,
    modifier: Modifier = Modifier,
    subtitle: String? = null,
    selected: Boolean = false,
    onTap: (() -> Unit)? = null,
) {
    val scheme = MaterialTheme.colorScheme
    val typography = MaterialTheme.typography
    val shape = NeptuneTheme.shape
    val feedback = NeptuneTheme.feedback
    val haptics: HapticFeedback = LocalHapticFeedback.current

    val clickableModifier = if (onTap != null) {
        Modifier.clickable {
            feedback.trigger(NptFeedbackCue.Tap, haptics)
            onTap()
        }
    } else {
        Modifier
    }

    Row(
        modifier = modifier
            .clip(shape.rMd)
            .background(scheme.surfaceContainerLowest)
            .then(clickableModifier)
            .border(
                width = if (selected) 2.dp else 1.dp,
                color = if (selected) scheme.primary else scheme.outlineVariant,
                shape = shape.rMd,
            )
            .defaultMinSize(minHeight = 64.dp)
            .padding(horizontal = 16.dp, vertical = 12.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Box(
            modifier = Modifier
                .size(40.dp)
                .clip(shape.rSm)
                .background(scheme.secondaryContainer),
            contentAlignment = Alignment.Center,
        ) {
            CompositionLocalProvider(LocalContentColor provides scheme.onSecondaryContainer) {
                Box(Modifier.size(20.dp), contentAlignment = Alignment.Center) { icon() }
            }
        }
        Spacer(Modifier.width(16.dp))
        Column(Modifier.weight(1f)) {
            Text(
                text = title,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis,
                style = typography.bodyLarge.copy(
                    color = scheme.onSurface,
                    fontWeight = FontWeight.W600,
                ),
            )
            if (subtitle != null) {
                Spacer(Modifier.height(2.dp))
                Text(
                    text = subtitle,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                    style = typography.bodyMedium.copy(color = scheme.onSurfaceVariant),
                )
            }
        }
        Spacer(Modifier.width(12.dp))
        // The trailing radio: fills primary with a check when selected, an
        // empty outlined circle otherwise.
        Box(
            modifier = Modifier
                .size(24.dp)
                .clip(shape.rFull)
                .then(if (selected) Modifier.background(scheme.primary) else Modifier)
                .border(
                    width = 2.dp,
                    color = if (selected) scheme.primary else scheme.outline,
                    shape = shape.rFull,
                ),
            contentAlignment = Alignment.Center,
        ) {
            if (selected) {
                Icon(
                    NptGlyphs.check,
                    contentDescription = null,
                    tint = scheme.onPrimary,
                    modifier = Modifier.size(16.dp),
                )
            }
        }
    }
}

/**
 * A selectable beneficiary tile: a circular 44dp avatar (provided [avatar]
 * slot or generated initials on `primaryContainer`), the [name], and a
 * masked [account] in tabular figures. When [selected] the tile lifts onto
 * `surfaceContainerHigh` with a primary ring and a trailing check-circle.
 * At least 56dp tall, whole tile tappable.
 *
 * Web counterpart: `<npt-beneficiary-tile>` · Flutter: `NeptuneBeneficiaryTile`.
 */
@Composable
public fun NeptuneBeneficiaryTile(
    name: String,
    modifier: Modifier = Modifier,
    account: String? = null,
    avatar: (@Composable () -> Unit)? = null,
    selected: Boolean = false,
    onTap: (() -> Unit)? = null,
) {
    val scheme = MaterialTheme.colorScheme
    val typography = MaterialTheme.typography
    val type = NeptuneTheme.type
    val shape = NeptuneTheme.shape
    val feedback = NeptuneTheme.feedback
    val haptics: HapticFeedback = LocalHapticFeedback.current
    val display = rememberNeptuneFontFamily(type.display)

    val accountStyle = NeptuneTheme.moneyStyle(base = typography.bodyMedium)
        .copy(color = scheme.onSurfaceVariant)

    val clickableModifier = if (onTap != null) {
        Modifier.clickable {
            feedback.trigger(NptFeedbackCue.Tap, haptics)
            onTap()
        }
    } else {
        Modifier
    }

    Row(
        modifier = modifier
            .clip(shape.rMd)
            .background(if (selected) scheme.surfaceContainerHigh else scheme.surface)
            .then(clickableModifier)
            .border(
                width = if (selected) 2.dp else 1.dp,
                color = if (selected) scheme.primary else scheme.outlineVariant,
                shape = shape.rMd,
            )
            .defaultMinSize(minHeight = 56.dp)
            .padding(horizontal = 12.dp, vertical = 12.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Box(Modifier.size(44.dp)) {
            if (avatar != null) {
                avatar()
            } else {
                Box(
                    modifier = Modifier
                        .fillMaxSize()
                        .clip(shape.rFull)
                        .background(scheme.primaryContainer),
                    contentAlignment = Alignment.Center,
                ) {
                    Text(
                        text = beneficiaryInitials(name),
                        style = typography.labelLarge.copy(
                            fontFamily = display,
                            color = scheme.onPrimaryContainer,
                            fontWeight = FontWeight.W600,
                        ),
                    )
                }
            }
        }
        Spacer(Modifier.width(16.dp))
        Column(Modifier.weight(1f)) {
            Text(
                text = name,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis,
                style = typography.bodyLarge.copy(color = scheme.onSurface),
            )
            if (account != null) {
                Spacer(Modifier.height(2.dp))
                Text(
                    text = NeptuneTheme.formatDigits(account),
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                    style = accountStyle,
                )
            }
        }
        if (selected) {
            Spacer(Modifier.width(12.dp))
            Icon(
                NptStatusGlyphs.successCheck,
                contentDescription = null,
                tint = scheme.primary,
                modifier = Modifier.size(22.dp),
            )
        }
    }
}

/** First letter of the first + last name parts, uppercased ('•' fallback) —
 * the Dart `_initials` recipe. */
private fun beneficiaryInitials(value: String): String {
    val parts = value.trim().split(Regex("\\s+")).filter { it.isNotEmpty() }
    if (parts.isEmpty()) return "•"
    val first = parts.first().first()
    val last = if (parts.size > 1) parts.last().first().toString() else ""
    val out = ("$first$last").uppercase()
    return out.ifEmpty { "•" }
}
